drivers/pktmem.c

```diff
@@ -15,80 +15,160 @@
+ *    BUF  - packet buffer
+ *    US   - unused space (could be empty)
+ *
+ *
+ *    PAGE_SIZE < pkt_sz + sizeof(pktmem_info_t):
+ *    +----+----+----+--     --+----+----+----+
+ *    |     PAGE1    |   ...   |     PAGEn    |
+ *    +----+----+----+--     --+----+----+----+
+ *    |                           |    |      |
+ *     --------------BUF---------- -US- -META-
+ *
+ *    PAGE_SIZE >= pkt_sz + sizeof(pktmem_info_t):
+ *    +----+----+----+--     ----+----+----+----+----+----+
+ *    |     BUF1     |   ...   |     BUFn     | US | META |
+ *    +----+----+----+--     ----+----+----+----+----+----+
+ *    |                                                   |
+ *     -----------------------PAGE------------------------
+ */
+static struct _pktmem_opts {
+	size_t alloc_sz;           /* sizeof full packet buffer, aligned to PAGE_SIZE */
+	size_t pkt_buf_sz;         /* sizeof space allocated for 1 packet buffer (DMA aligned) */
+	size_t pkt_buf_cnt;        /* number of packet buffers in a single pktmem */
+	size_t pktmem_info_offset; /* metadata offset from the pktmem start */
+	unsigned free_mask_full;   /* bitmask representing all packet buffers in a single pktmem */
+} pktmem_opts = { 0 };
 
 
-typedef struct _buf_page_head {
+typedef struct _pktmem_info {
 	unsigned free_mask;
-	struct _buf_page_head *next;
-	struct _buf_page_head *prev;
-} buf_page_head_t;
+	struct _pktmem_info *next;
+	struct _pktmem_info *prev;
+} pktmem_info_t;
+
 
+struct pktmem_pbuf_custom {
+	struct pbuf_custom pbuf_custom;
+	/* keep a pointer to the start of pktmem to free it correctly
+	 * this is necessary on targets with PAGE_SIZE < pkt_sz + sizeof(pktmem_info_t)
+	 * as we cannot determine the start of pktmem from a given payload pointer
+	 */
+	uint8_t _padding[16];
+	void *pktmem_start;
+};
 
-static buf_page_head_t pkt_buf_lru = { .next = &pkt_buf_lru, .prev = &pkt_buf_lru };
+
+static pktmem_info_t pktmem_lru = { .next = &pktmem_lru, .prev = &pktmem_lru };
 static unsigned pkt_bufs_free;
 
 
-static void net_listAdd(buf_page_head_t *ph, buf_page_head_t *after)
+static void net_listAdd(pktmem_info_t *info, pktmem_info_t *after)
 {
-	buf_page_head_t *next;
+	pktmem_info_t *next;
 
-	next = ph->next = after->next;
-	ph->prev = after;
-	next->prev = ph;
-	after->next = ph;
+	next = info->next = after->next;
+	info->prev = after;
+	next->prev = info;
+	after->next = info;
 }
 
 
-static void net_listDel(buf_page_head_t *ph)
+static void net_listDel(pktmem_info_t *info)
 {
-	buf_page_head_t *prev, *next;
+	pktmem_info_t *prev, *next;
 
-	prev = ph->prev;
-	next = ph->next;
+	prev = info->prev;
+	next = info->next;
 	prev->next = next;
 	next->prev = prev;
 }
 
 
-static void net_freePktBuf(void *p)
+static inline bool net_isPktMemInitialized(void)
+{
+	return pktmem_opts.alloc_sz != 0;
+}
+
+
+int net_initPktMem(size_t pkt_buf_max_sz)
+{
+	if (net_isPktMemInitialized()) {
+		return 0;
+	}
+
+	if (pkt_buf_max_sz == 0) {
+		return -EINVAL;
+	}
+
```

Comment:
Please return `-EINVAL` if `sizeof(pktmem_info_t) % __alignof__(pktmem_info_t) != 0` to ensure that `pktmem_info_t` structure is properly aligned.